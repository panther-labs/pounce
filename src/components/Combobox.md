A simple Combobox can be thought of as a typical `<select>` component. Whenerever you would
use a normal select, you should now pass the `<Combobox>` component.

A thing to remember is that all the comboboxes have a controlled state when
it comes to the selected items. Thus, they do not maintain their own internal state
but expect you to pass them explicitly the selected items.

You get notified of changes to the selections through the `onChange` prop.

---

A Combobox can be basic:

```jsx harmony
import React from 'react';

const Example1 = () => {
  const [selectedItems, updateSelectedItems] = React.useState(null);
  return (
    <Combobox
      label="Choose a car manufacturer"
      items={[
        { text: 'Toyota', value: 'toyota' },
        { text: 'Ford', value: 'ford' },
        { text: 'Chevrolet', value: 'chevrolet' },
        { text: 'BMW', value: 'bmw' },
        { text: 'Mercedes', value: 'mercedes' },
        { text: 'Hammer', value: 'hammer' },
        { text: 'Dodge', value: 'dodge' },
        { text: 'Audi', value: 'audi' },
      ]}
      onChange={updateSelectedItems}
      selectedItem={selectedItems}
      inputProps={{ placeholder: 'Search for a manufacturer' }}
      rootProps={{ width: 400 }}
    />
  );
};

<Example1 />;
```

A Combobox can have search functionality:

```jsx harmony
import React from 'react';

const Example2 = () => {
  const [selectedItems, updateSelectedItems] = React.useState(null);
  return (
    <Combobox
      searchable
      label="Choose a car manufacturer"
      items={[
        { text: 'Toyota', value: 'toyota' },
        { text: 'Ford', value: 'ford' },
        { text: 'Chevrolet', value: 'chevrolet' },
        { text: 'BMW', value: 'bmw' },
        { text: 'Mercedes', value: 'mercedes' },
        { text: 'Hammer', value: 'hammer' },
        { text: 'Dodge', value: 'dodge' },
        { text: 'Audi', value: 'audi' },
      ]}
      onChange={updateSelectedItems}
      selectedItem={selectedItems}
      inputProps={{ placeholder: 'Search for a manufacturer' }}
      rootProps={{ width: 400 }}
    />
  );
};

<Example2 />;
```

A Combobox can accepts multiple values:

```jsx harmony
import React from 'react';

const Example3 = () => {
  const [selectedItems, updateSelectedItems] = React.useState([]);
  return (
    <Combobox
      label="Choose a car manufacturer"
      items={[
        { text: 'Toyota', value: 'toyota' },
        { text: 'Ford', value: 'ford' },
        { text: 'Chevrolet', value: 'chevrolet' },
        { text: 'BMW', value: 'bmw' },
        { text: 'Mercedes', value: 'mercedes' },
        { text: 'Hammer', value: 'hammer' },
        { text: 'Dodge', value: 'dodge' },
        { text: 'Audi', value: 'audi' },
      ]}
      onChange={updateSelectedItems}
      selectedItem={selectedItems}
      inputProps={{ placeholder: 'Search for a manufacturer' }}
      rootProps={{ width: 400 }}
    />
  );
};

<Example3 />;
```

A Combobox can accepts multiple values and have search functionality:

```jsx harmony
import React from 'react';

const Example4 = () => {
  const [selectedItems, updateSelectedItems] = React.useState([]);
  return (
    <Combobox
      searchable
      label="Choose a car manufacturer"
      items={[
        { text: 'Toyota', value: 'toyota' },
        { text: 'Ford', value: 'ford' },
        { text: 'Chevrolet', value: 'chevrolet' },
        { text: 'BMW', value: 'bmw' },
        { text: 'Mercedes', value: 'mercedes' },
        { text: 'Hammer', value: 'hammer' },
        { text: 'Dodge', value: 'dodge' },
        { text: 'Audi', value: 'audi' },
      ]}
      onChange={updateSelectedItems}
      selectedItem={selectedItems}
      inputProps={{ placeholder: 'Search for a manufacturer' }}
      rootProps={{ width: 400 }}
    />
  );
};

<Example4 />;
```
