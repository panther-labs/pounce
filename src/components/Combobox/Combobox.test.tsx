import { renderWithTheme, fireClickAndMouseEvents, waitMs } from 'test-utils';
import React from 'react';
import Combobox, { ComboboxProps } from './index';
import userEvent from '@testing-library/user-event';

type Item = { value: string; category: string };

const items: Item[] = [
  { value: 'Toyota', category: 'Normal' },
  { value: 'Ford', category: 'Normal' },
  { value: 'Chevrolet', category: 'Luxury' },
];

const ControlledCombobox: React.FC<Partial<ComboboxProps<Item>>> = props => {
  const [selectedManufacturer, updateSelectedManufacturer] = React.useState<Item | null>(null);
  return (
    <Combobox
      label="Choose a car manufacturer"
      onChange={updateSelectedManufacturer}
      value={selectedManufacturer}
      placeholder="Select manufacturer"
      items={items}
      itemToString={item => item.value}
      {...props}
    />
  );
};

describe('Combobox', () => {
  it('renders', () => {
    const { container, getByText, getByPlaceholderText } = renderWithTheme(<ControlledCombobox />);
    expect(container).toMatchSnapshot();

    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturer'));
    expect(container).toMatchSnapshot();

    fireClickAndMouseEvents(getByText('Toyota'));
    expect(container).toMatchSnapshot();
  });

  it('doesn\'t render a "clear selection" button when no option is selected', () => {
    const { queryByText, getByPlaceholderText } = renderWithTheme(<ControlledCombobox />);

    fireClickAndMouseEvents(getByPlaceholderText('Select manufacturer'));
    expect(queryByText('Clear Selection')).not.toBeInTheDocument();
  });

  it('clears selected value when the "clear selection" button is clicked', () => {
    const { getByText, getByPlaceholderText } = renderWithTheme(<ControlledCombobox />);

    const selectionInput = getByPlaceholderText('Select manufacturer');
    const selectionValue = 'Toyota';

    expect(selectionInput).not.toHaveValue();
    fireClickAndMouseEvents(selectionInput);
    fireClickAndMouseEvents(getByText(selectionValue));
    expect(selectionInput).toHaveValue(selectionValue);

    const clearSelectionButton = getByText(
      (_, { tagName, textContent }) => tagName === 'BUTTON' && textContent === 'Clear Selection'
    );

    fireClickAndMouseEvents(clearSelectionButton);
    expect(selectionInput).not.toHaveValue();
  });

  it('fires onblur event when the input focus is lost', async () => {
    const mock = jest.fn();
    const { getByPlaceholderText } = renderWithTheme(
      <>
        <ControlledCombobox onBlur={mock} />
        <input placeholder="another input" />
      </>
    );
    userEvent.click(getByPlaceholderText('Select manufacturer'));
    userEvent.click(getByPlaceholderText('another input'));
    expect(mock).toHaveBeenCalledTimes(1);
  });

  it('toggles the dropdown menu by clicking the caret button', async () => {
    const { getByText, getByPlaceholderText, getByRole, queryByText } = renderWithTheme(
      <ControlledCombobox searchable />
    );
    const toggleMenuButton = getByRole('button', { name: 'Toggle Menu' });
    userEvent.click(getByPlaceholderText('Select manufacturer'));
    expect(getByText('Toyota')).toBeInTheDocument();
    fireClickAndMouseEvents(getByText('Toyota'));
    await waitMs(300);
    // Check that menu is closed after selection
    expect(queryByText('Ford')).not.toBeInTheDocument();

    // Click toggle menu button and expect menu to open
    fireClickAndMouseEvents(toggleMenuButton);
    await waitMs(300);
    expect(queryByText('Ford')).toBeInTheDocument();

    // Click toggle menu button again and expect menu to close
    fireClickAndMouseEvents(toggleMenuButton);
    await waitMs(300);
    expect(queryByText('Ford')).not.toBeInTheDocument();
  });
});
