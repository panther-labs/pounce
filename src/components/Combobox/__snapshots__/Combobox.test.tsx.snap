// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox renders 1`] = `
.pounce-6 {
  position: relative;
}

.pounce-2 {
  min-height: 47px;
  position: relative;
  border: 1px solid;
  -webkit-transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  background-color: transparent;
  border-radius: 4px;
  border-color: #3E516B;
}

.pounce-2:hover {
  border-color: #0081C5;
}

.pounce-2:focus-within {
  border-color: #0081C5;
}

.pounce-2:focus-within span {
  opacity: 1;
  visibility: visible;
}

.pounce-2:focus-within label {
  font-weight: 500;
  -webkit-transform: translate(6px, 4px) scale(0.65);
  -moz-transform: translate(6px, 4px) scale(0.65);
  -ms-transform: translate(6px, 4px) scale(0.65);
  transform: translate(6px, 4px) scale(0.65);
}

.pounce-0 {
  width: 100%;
  height: 100%;
  padding-left: 16px;
  padding-right: 32px;
  padding-top: 20px;
  padding-bottom: 8px;
  position: relative;
  color: #F6F6F6;
  font-size: 0.875rem;
  font-weight: 500;
  background-color: transparent;
  border: 0;
  cursor: pointer;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.pounce-0::-webkit-input-placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::-moz-placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0:-ms-input-placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}



.pounce-0:focus::-webkit-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::-moz-placeholder {
  opacity: 0.4;
}

.pounce-0:focus:-ms-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::placeholder {
  opacity: 0.4;
}

.pounce-0:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 30px #192231 inset;
  -webkit-text-fill-color: #F6F6F6;
  border-radius: 4px;
}

.pounce-1 {
  pointer-events: none;
  font-size: 0.875rem;
  padding-left: 16px;
  padding-right: 16px;
  color: #BDBDBD;
  top: 0;
  left: 0;
  position: absolute;
  transform-origin: center left;
  -webkit-transform: translate(0, 14px) scale(1);
  -moz-transform: translate(0, 14px) scale(1);
  -ms-transform: translate(0, 14px) scale(1);
  transform: translate(0, 14px) scale(1);
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 400;
}

.pounce-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 1;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 8px;
  pointer-events: none;
}

.pounce-4 {
  outline: none;
  line-height: 1;
  border-radius: 4px;
  height: 32px;
  width: 32px;
  cursor: pointer;
  color: #F6F6F6;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
}

.pounce-4:focus {
  background-color: rgba(255,255,255,0.1);
}

.pounce-3 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  color: #FFFFFF;
}



<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    class="pounce-6"
    role="combobox"
  >
    <div
      class="pounce-6"
    >
      <div
        aria-disabled="false"
        class="pounce-2"
      >
        <input
          aria-autocomplete="list"
          aria-disabled="false"
          aria-invalid="false"
          aria-labelledby="downshift-0-label"
          aria-readonly="true"
          autocomplete="off"
          class="pounce-0"
          id="downshift-0-input"
          placeholder="Select manufacturer"
          readonly=""
          type="text"
          value=""
        />
        <label
          class="pounce-1"
          for="downshift-0-input"
          id="downshift-0-label"
        >
          Choose a car manufacturer
        </label>
      </div>
      <div
        class="pounce-5"
      >
        <button
          aria-label="Toggle Menu"
          aria-pressed="false"
          class="pounce-4"
          tabindex="-1"
          type="button"
        >
          <svg
            class="pounce-3"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10L12 15L17 10H7Z"
            />
          </svg>
        </button>
      </div>
    </div>
    <ul
      aria-labelledby="downshift-0-label"
      class="pounce-7"
      id="downshift-0-menu"
      role="listbox"
    />
  </div>
</div>
`;

exports[`Combobox renders 2`] = `
.pounce-6 {
  position: relative;
}

.pounce-0 {
  width: 100%;
  height: 100%;
  padding-left: 16px;
  padding-right: 32px;
  padding-top: 20px;
  padding-bottom: 8px;
  position: relative;
  color: #F6F6F6;
  font-size: 0.875rem;
  font-weight: 500;
  background-color: transparent;
  border: 0;
  cursor: pointer;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.pounce-0::-webkit-input-placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::-moz-placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0:-ms-input-placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}



.pounce-0:focus::-webkit-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::-moz-placeholder {
  opacity: 0.4;
}

.pounce-0:focus:-ms-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::placeholder {
  opacity: 0.4;
}

.pounce-0:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 30px #192231 inset;
  -webkit-text-fill-color: #F6F6F6;
  border-radius: 4px;
}

.pounce-1 {
  pointer-events: none;
  font-size: 0.875rem;
  padding-left: 16px;
  padding-right: 16px;
  color: #BDBDBD;
  top: 0;
  left: 0;
  position: absolute;
  transform-origin: center left;
  -webkit-transform: translate(0, 14px) scale(1);
  -moz-transform: translate(0, 14px) scale(1);
  -ms-transform: translate(0, 14px) scale(1);
  transform: translate(0, 14px) scale(1);
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 400;
}

.pounce-4 {
  outline: none;
  line-height: 1;
  border-radius: 4px;
  height: 32px;
  width: 32px;
  cursor: pointer;
  color: #F6F6F6;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
}

.pounce-4:focus {
  background-color: rgba(255,255,255,0.1);
}

.pounce-3 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  color: #FFFFFF;
}



.pounce-2 {
  min-height: 47px;
  position: relative;
  border: 1px solid;
  -webkit-transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  background-color: #23344E;
  border-radius: 4px;
  border-color: #23344E;
}

.pounce-2:hover {
  border-color: #0081C5;
}

.pounce-2:focus-within {
  border-color: #0081C5;
}

.pounce-2:focus-within span {
  opacity: 1;
  visibility: visible;
}

.pounce-2:focus-within label {
  font-weight: 500;
  -webkit-transform: translate(6px, 4px) scale(0.65);
  -moz-transform: translate(6px, 4px) scale(0.65);
  -ms-transform: translate(6px, 4px) scale(0.65);
  transform: translate(6px, 4px) scale(0.65);
}

.pounce-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 1;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 8px;
  pointer-events: auto;
}

.pounce-11 {
  margin-top: 6px;
  border: 1px solid;
  border-color: #0081C5;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
  background-color: #3E516B;
  z-index: 10;
  position: absolute;
  min-width: 100%;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  max-width: 40vw;
  max-height: 300px;
  overflow: auto;
}

.pounce-8 {
  cursor: pointer;
  font-size: 0.875rem;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-right: 16px;
  padding-left: 16px;
  -webkit-transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  position: relative;
  background-color: #3E516B;
  word-break: break-all;
}

.pounce-8[aria-selected=true],
[data-selected]>.pounce-8 {
  background-color: #23344E;
}

<div>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    aria-owns="downshift-0-menu"
    class="pounce-6"
    role="combobox"
  >
    <div
      class="pounce-6"
    >
      <div
        aria-disabled="false"
        class="pounce-2"
      >
        <input
          aria-autocomplete="list"
          aria-controls="downshift-0-menu"
          aria-disabled="false"
          aria-invalid="false"
          aria-labelledby="downshift-0-label"
          aria-readonly="true"
          autocomplete="off"
          class="pounce-0"
          id="downshift-0-input"
          placeholder="Select manufacturer"
          readonly=""
          type="text"
          value=""
        />
        <label
          class="pounce-1"
          for="downshift-0-input"
          id="downshift-0-label"
        >
          Choose a car manufacturer
        </label>
      </div>
      <div
        class="pounce-5"
      >
        <button
          aria-label="Toggle Menu"
          aria-pressed="false"
          class="pounce-4"
          tabindex="-1"
          type="button"
        >
          <svg
            class="pounce-3"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 14L12 9L17 14H7Z"
            />
          </svg>
        </button>
      </div>
    </div>
    <ul
      aria-labelledby="downshift-0-label"
      class="pounce-7"
      id="downshift-0-menu"
      role="listbox"
    />
    <ul
      aria-labelledby="downshift-0-label"
      class="pounce-11"
      id="downshift-0-menu"
      role="listbox"
      style="transform: scale(1, 1); opacity: 1;"
    >
      <li
        aria-disabled="false"
        aria-selected="false"
        class="pounce-8"
        id="downshift-0-item-0"
        role="option"
      >
        Toyota
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="pounce-8"
        id="downshift-0-item-1"
        role="option"
      >
        Ford
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="pounce-8"
        id="downshift-0-item-2"
        role="option"
      >
        Chevrolet
      </li>
    </ul>
  </div>
</div>
`;

exports[`Combobox renders 3`] = `
.pounce-6 {
  position: relative;
}

.pounce-2 {
  min-height: 47px;
  position: relative;
  border: 1px solid;
  -webkit-transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  background-color: transparent;
  border-radius: 4px;
  border-color: #3E516B;
}

.pounce-2:hover {
  border-color: #0081C5;
}

.pounce-2:focus-within {
  border-color: #0081C5;
}

.pounce-2:focus-within span {
  opacity: 1;
  visibility: visible;
}

.pounce-2:focus-within label {
  font-weight: 500;
  -webkit-transform: translate(6px, 4px) scale(0.65);
  -moz-transform: translate(6px, 4px) scale(0.65);
  -ms-transform: translate(6px, 4px) scale(0.65);
  transform: translate(6px, 4px) scale(0.65);
}

.pounce-0 {
  width: 100%;
  height: 100%;
  padding-left: 16px;
  padding-right: 32px;
  padding-top: 20px;
  padding-bottom: 8px;
  position: relative;
  color: #F6F6F6;
  font-size: 0.875rem;
  font-weight: 500;
  background-color: transparent;
  border: 0;
  cursor: pointer;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.pounce-0::-webkit-input-placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::-moz-placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0:-ms-input-placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-0::placeholder {
  opacity: 0;
  color: #F6F6F6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}



.pounce-0:focus::-webkit-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::-moz-placeholder {
  opacity: 0.4;
}

.pounce-0:focus:-ms-input-placeholder {
  opacity: 0.4;
}

.pounce-0:focus::placeholder {
  opacity: 0.4;
}

.pounce-0:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 30px #192231 inset;
  -webkit-text-fill-color: #F6F6F6;
  border-radius: 4px;
}

.pounce-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 1;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 8px;
  pointer-events: none;
}

.pounce-4 {
  outline: none;
  line-height: 1;
  border-radius: 4px;
  height: 32px;
  width: 32px;
  cursor: pointer;
  color: #F6F6F6;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
}

.pounce-4:focus {
  background-color: rgba(255,255,255,0.1);
}

.pounce-3 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  color: #FFFFFF;
}



.pounce-15 {
  margin-top: 6px;
  border: 1px solid;
  border-color: #0081C5;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
  background-color: #3E516B;
  z-index: 10;
  position: absolute;
  min-width: 100%;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  max-width: 40vw;
  max-height: 300px;
  overflow: auto;
}

.pounce-13 {
  cursor: pointer;
  font-size: 0.875rem;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-right: 16px;
  padding-left: 16px;
  -webkit-transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  position: relative;
  background-color: #3E516B;
  word-break: break-all;
}

.pounce-13[aria-selected=true],
[data-selected]>.pounce-13 {
  background-color: #23344E;
}

.pounce-1 {
  pointer-events: none;
  font-size: 0.875rem;
  padding-left: 16px;
  padding-right: 16px;
  color: #BDBDBD;
  top: 0;
  left: 0;
  position: absolute;
  transform-origin: center left;
  -webkit-transform: translate(6px, 4px) scale(0.65);
  -moz-transform: translate(6px, 4px) scale(0.65);
  -ms-transform: translate(6px, 4px) scale(0.65);
  transform: translate(6px, 4px) scale(0.65);
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.pounce-11 {
  background-color: #30415C;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1;
}

.pounce-10 {
  width: 100%;
  font-size: 0.625rem;
  color: #56E3C9;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
}

.pounce-10:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.pounce-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 6px;
  padding-bottom: 6px;
  padding-left: 16px;
  padding-right: 16px;
}

.pounce-9>*:not(:last-child) {
  margin-right: 6px;
}

.pounce-7 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 16px;
  height: 16px;
  color: currentColor;
}

.pounce-12 {
  cursor: pointer;
  font-size: 0.875rem;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-right: 16px;
  padding-left: 16px;
  -webkit-transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  position: relative;
  background-color: #1D2A3F;
  word-break: break-all;
}

.pounce-12[aria-selected=true],
[data-selected]>.pounce-12 {
  background-color: #23344E;
}

<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    class="pounce-6"
    role="combobox"
  >
    <div
      class="pounce-6"
    >
      <div
        aria-disabled="false"
        class="pounce-2"
      >
        <input
          aria-autocomplete="list"
          aria-disabled="false"
          aria-invalid="false"
          aria-labelledby="downshift-0-label"
          aria-readonly="true"
          autocomplete="off"
          class="pounce-0"
          id="downshift-0-input"
          placeholder="Select manufacturer"
          readonly=""
          type="text"
          value="Toyota"
        />
        <label
          class="pounce-1"
          for="downshift-0-input"
          id="downshift-0-label"
        >
          Choose a car manufacturer
        </label>
      </div>
      <div
        class="pounce-5"
      >
        <button
          aria-label="Toggle Menu"
          aria-pressed="false"
          class="pounce-4"
          tabindex="-1"
          type="button"
        >
          <svg
            class="pounce-3"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10L12 15L17 10H7Z"
            />
          </svg>
        </button>
      </div>
    </div>
    <ul
      aria-labelledby="downshift-0-label"
      class="pounce-15"
      id="downshift-0-menu"
      role="listbox"
      style="transform: scale(0.9, 0.9); opacity: 0;"
    >
      <li
        class="pounce-11"
      >
        <button
          class="pounce-10"
          type="button"
        >
          <span
            class="pounce-9"
          >
            <svg
              class="pounce-7"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 2c5.523 0 10 4.477 10 10s-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2zm2.828 5.757L12 10.587l-2.828-2.83-1.415 1.415L10.587 12l-2.83 2.828 1.415 1.415L12 13.414l2.828 2.829 1.415-1.415L13.414 12l2.829-2.828-1.415-1.415z"
              />
            </svg>
            <span
              class="pounce-8"
            >
              Clear Selection
            </span>
          </span>
        </button>
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="pounce-12"
        id="downshift-0-item-0"
        role="option"
      >
        Toyota
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="pounce-13"
        id="downshift-0-item-1"
        role="option"
      >
        Ford
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="pounce-13"
        id="downshift-0-item-2"
        role="option"
      >
        Chevrolet
      </li>
    </ul>
    <ul
      aria-labelledby="downshift-0-label"
      class="pounce-8"
      id="downshift-0-menu"
      role="listbox"
    />
    <ul
      aria-labelledby="downshift-0-label"
      class="pounce-8"
      id="downshift-0-menu"
      role="listbox"
    />
  </div>
</div>
`;
